### PR TITLE
Sprint 2

### PR DESCRIPTION
**Pull request by 23mafi**
adds submission of sprint 2

### CODE REVIEW

File: src/lib/CMakeLists.txt
Comment (T4rikA) [inline src/lib/CMakeLists.txt:1]: Why is this changed in this PR?

File: src/lib/storage/chunk.cpp
Comment (T4rikA) [inline src/lib/storage/chunk.cpp:1]: If this is the result from the feedback of the first sprint, this should have been handled in the PR for the first sprint. Maybe move it to a new PR now.

File: src/lib/storage/chunk.cpp [hunk 1]
```diff
@@ -9,20 +9,19 @@ namespace opossum {
...
-  } else if (const auto concrete_float = std::dynamic_pointer_cast<ValueSegment<float>>(segment)) {
-    concrete_float->append(value);
-  } else if (const auto concrete_double = std::dynamic_pointer_cast<ValueSegment<double>>(segment)) {
-    concrete_double->append(value);
-  } else if (const auto concrete_string = std::dynamic_pointer_cast<ValueSegment<std::string>>(segment)) {
-    concrete_string->append(value);
-  } else {
-    Fail("Datatype not implemented.");
+// Test all possible ValueSegment variants.
+#define TRY_APPEND_TYPE(Type)                                                         \
```
Comment (T4rikA) [inline src/lib/storage/chunk.cpp:13]: Why do you prefer the macro in comparison to the solution presented in the lecture (in terms of readability and maintainability)?

File: src/lib/storage/dictionary_segment.cpp [hunk 2]
```diff
@@ -1,95 +1,159 @@
 #include "dictionary_segment.hpp"
+#include <map>
```
Comment (T4rikA) [inline src/lib/storage/dictionary_segment.cpp]: Remove unused include

File: src/lib/storage/dictionary_segment.cpp [hunk 3]
```diff
@@ -1,95 +1,159 @@
...
+template <typename T>
+void DictionarySegment<T>::_compress(const std::shared_ptr<AbstractSegment>& abstract_segment) {
+  _create_dictionary(abstract_segment);
+  _create_attribute_vector(abstract_segment);
+}
+
+template <typename T>
+void DictionarySegment<T>::_create_dictionary(const std::shared_ptr<AbstractSegment>& abstract_segment) {
+  auto unique_values = std::set<T>();
+  const auto size = abstract_segment->size();
```
Comment (T4rikA) [inline src/lib/storage/dictionary_segment.cpp]: Maybe choose a more descriptive variable name (e.g. segment_size)

File: src/lib/storage/dictionary_segment.cpp [hunk 4]
```diff
@@ -1,95 +1,159 @@
...
+
+  // Insert all values into an ordered map.
```
Comment (T4rikA) [inline src/lib/storage/dictionary_segment.cpp]:
```suggestion
  // Insert all values into a set.
```

File: src/lib/storage/dictionary_segment.cpp [hunk 5]
```diff
@@ -1,95 +1,159 @@
...
+  for (auto index = ChunkOffset{0}; index < size; ++index) {
+    const auto variant = abstract_segment->operator[](index);
+    if (variant_is_null(variant)) {
+      continue;
+    }
+    const auto typed_value = type_cast<T>(variant);
+    unique_values.insert(typed_value);
```
Comment (T4rikA) [inline src/lib/storage/dictionary_segment.cpp:33]:
```suggestion
    if (!variant_is_null(variant)) {
        const auto typed_value = type_cast<T>(variant);
        unique_values.insert(typed_value);
    }
```

File: src/lib/storage/dictionary_segment.cpp [hunk 6]
```diff
@@ -1,95 +1,159 @@
...
+  }
+
+  // Create the sorted dictionary from ordered map.
```
Comment (T4rikA) [inline src/lib/storage/dictionary_segment.cpp]:
```suggestion
  // Create the sorted dictionary from set.
```

File: src/lib/storage/dictionary_segment.cpp [hunk 7]
```diff
@@ -1,95 +1,159 @@
...
+  _dictionary.reserve(unique_values.size());
+  for (const auto& value : unique_values) {
+    _dictionary.push_back(value);
+  }
+}
+
+template <typename T>
+void DictionarySegment<T>::_create_attribute_vector(const std::shared_ptr<AbstractSegment>& abstract_segment) {
+  auto attribute_list = std::vector<ValueID>();
+  const auto size = abstract_segment->size();
```
Comment (T4rikA) [inline src/lib/storage/dictionary_segment.cpp]: Same as above

File: src/lib/storage/dictionary_segment.cpp [hunk 8]
```diff
@@ -1,95 +1,159 @@
...
+  attribute_list.reserve(size);
+
+  // Iterate over values again to get index and save that in the attribute_list.
+  for (auto index = ChunkOffset{0}; index < size; ++index) {
+    const auto variant = abstract_segment->operator[](index);
+    if (variant_is_null(variant)) {
+      attribute_list.push_back(null_value_id());
+      continue;
```
Comment (T4rikA) [inline src/lib/storage/dictionary_segment.cpp:54]: Use else instead

File: src/lib/storage/dictionary_segment.cpp [hunk 9]
```diff
@@ -1,95 +1,159 @@
...
+    }
+    const auto typed_value = type_cast<T>(variant);
+    const auto it = std::find(_dictionary.begin(), _dictionary.end(), typed_value);
```
Comment (T4rikA) [inline src/lib/storage/dictionary_segment.cpp]: Here you search every value (n) in an unordered vector (n) which is O(n^2), this can be done faster by sorting the vector once or using a sorted data structure.

File: src/lib/storage/dictionary_segment.cpp [hunk 10]
```diff
@@ -1,95 +1,159 @@
...
+    DebugAssert(it != _dictionary.end(), "Inserted value not in the set of unique values.");
+    // + 1 to account for null_value_id().
+    const auto dictionary_index = std::distance(_dictionary.begin(), it) + 1;
```
Comment (T4rikA) [inline src/lib/storage/dictionary_segment.cpp]: This would be easier with a map (value -> id), would also fix the runtime problem

File: src/lib/storage/dictionary_segment.cpp [hunk 11]
```diff
@@ -1,95 +1,159 @@
...
+  }
+  return NULL_VALUE;
 }
 
 template <typename T>
 T DictionarySegment<T>::get(const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  const auto optional = get_typed_value(chunk_offset);
+  Assert(optional, "Value is NULL.");
```
Comment (T4rikA) [inline src/lib/storage/dictionary_segment.cpp]: Give a more descriptive error, e.g. include the `chunk_offset`

File: src/lib/storage/dictionary_segment.cpp
Comment (T4rikA) [inline src/lib/storage/dictionary_segment.cpp:1]: - You have to check if the segment is nullable, your code should behave different for each case
- Use `dictionary()` instead of `_dictionary` (same for attribute vector)

File: src/lib/storage/dictionary_segment.cpp [hunk 12]
```diff
@@ -1,95 +1,159 @@
...
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return ValueID{0};
 }
 
 template <typename T>
 const T DictionarySegment<T>::value_of_value_id(const ValueID value_id) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  DebugAssert(value_id != null_value_id(), "ValueID " + std::to_string(value_id) + " is reserved for NULL values.");
```
Comment (T4rikA) [inline src/lib/storage/dictionary_segment.cpp]: This should only be this way if segment is nullable

File: src/lib/storage/dictionary_segment.cpp [hunk 13]
```diff
@@ -1,95 +1,159 @@
...
+  // null_value_id is 0, so every other index is off by one.
+  const auto value_id_without_null = ValueID{value_id - 1};
```
Comment (T4rikA) [inline src/lib/storage/dictionary_segment.cpp]: Depends on whether the segment is nullable

File: src/lib/storage/dictionary_segment.cpp [hunk 14]
```diff
@@ -1,95 +1,159 @@
...
 ChunkOffset DictionarySegment<T>::size() const {
-  // Implementation goes here
-  return ChunkOffset{};
+  return ChunkOffset(attribute_vector()->size());
 }
 
 template <typename T>
 size_t DictionarySegment<T>::estimate_memory_usage() const {
-  return size_t{};
+  return sizeof(T) * _dictionary.capacity() + attribute_vector()->width() * attribute_vector()->size();
```
Comment (T4rikA) [inline src/lib/storage/dictionary_segment.cpp]: You use the capacity for the dictionary and the size for the attribute vector, is there a specific reason for that?

File: src/lib/storage/fixed_width_integer_vector.cpp [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+  return sizeof(T);
+}
+
+template <typename T>
+ValueID FixedWidthIntegerVector<T>::get(const size_t index) const {
+  Assert(index < size(), "Index " + std::to_string(index) + " is out of range.");
+  return ValueID{_value_ids[index]};
+}
+
+std::shared_ptr<AbstractAttributeVector> compress_attribute_vector(const std::vector<ValueID>& attribute_list) {
```
Comment (T4rikA) [inline src/lib/storage/fixed_width_integer_vector.cpp]: Why is this method here, we tried to keep the interface as small as possible.
Furthermore we don't think this should be the responsibility of the FixedWidthIntegerVector. This class should not hold specific compress functionality

File: src/lib/storage/fixed_width_integer_vector.cpp [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+  const auto max_element = std::max_element(attribute_list.begin(), attribute_list.end());
+  if (max_element == attribute_list.end()) {
+    return std::make_shared<FixedWidthIntegerVector<uint8_t>>();
+  }
+  // +1 due to null value
+  const auto total_number_of_values = static_cast<size_t>(*max_element) + 1;
```
Comment (T4rikA) [inline src/lib/storage/fixed_width_integer_vector.cpp]: One example of the vector holding a lot of information it shouldn't have

File: src/lib/storage/fixed_width_integer_vector.cpp (hunk 15 above)
Comment (T4rikA) [inline src/lib/storage/fixed_width_integer_vector.cpp]:
```suggestion
std::shared_ptr<AbstractAttributeVector> compress_attribute_vector(const std::vector<ValueID>& attributes) {
```
We find this confusing as it's not a list

File: src/lib/storage/fixed_width_integer_vector.cpp (hunk 16 above)
Comment (T4rikA) [reply src/lib/storage/fixed_width_integer_vector.cpp]: You can use the `std::bit_width` function

File: src/lib/storage/fixed_width_integer_vector.hpp [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+class FixedWidthIntegerVector : public AbstractAttributeVector {
+ public:
+  FixedWidthIntegerVector() = default;
+  explicit FixedWidthIntegerVector(size_t size);
+  explicit FixedWidthIntegerVector(const std::vector<ValueID>& values);
+  ~FixedWidthIntegerVector() override = default;
+
+  // We need to explicitly set the move constructor to default when we overwrite the copy constructor.
+  FixedWidthIntegerVector(FixedWidthIntegerVector&&) noexcept = default;
+  FixedWidthIntegerVector& operator=(FixedWidthIntegerVector&&) noexcept = default;
```
Comment (T4rikA) [inline src/lib/storage/fixed_width_integer_vector.hpp]: We don't think you actually need this, but are not 100% sure

File: src/lib/storage/table.cpp
Comment (T4rikA) [inline src/lib/storage/table.cpp:1]: Please check if all changes in this file should be changed in this PR

File: src/lib/storage/value_segment.cpp
Comment (T4rikA) [inline src/lib/storage/value_segment.cpp:1]: Should be changed in a different PR

File: src/test/storage/chunk_test.cpp
Comment (T4rikA) [inline src/test/storage/chunk_test.cpp:1]: Should be changed in a different PR

File: src/test/storage/table_test.cpp
Comment (T4rikA) [inline src/test/storage/table_test.cpp:1]: Should be changed in a different PR

File: src/test/storage/value_segment_test.cpp
Comment (T4rikA) [inline src/test/storage/value_segment_test.cpp:1]: Should be changed in a different PR

**Review by T4rikA — CHANGES_REQUESTED** [review]
We really liked your test cases!

We mentioned for some files, that they should be changed in a different PR. We are not sure how the teaching team wants us to implement the review, but for us it only makes sense to do this directly in the corresponding PR and not in the PR for the next sprint.

You have to check for the compression whether a segment can be nullable or not, this impacts how the compression must be done.

Furthermore check the responsibility of the FixedWidthIntegerVector, we think it should  not hold information about the compression

Paula and Tarik
Team 3

File: src/lib/storage/table.cpp [hunk 18]
```diff
@@ -106,8 +111,59 @@ std::shared_ptr<const Chunk> Table::get_chunk(ChunkID chunk_id) const {
...
+    thread.detach();
+  }
+
+  // Create new empty chunk and append segments to it.
+  auto compressed_chunk = std::make_shared<Chunk>();
+  for (auto& future : futures) {
+    future.wait();
+    const auto compressed_segment = future.get();
+    compressed_chunk->add_segment(compressed_segment);
+  }
```
Comment (SanJSp) [inline src/lib/storage/table.cpp:167]: TIL what futures are 😁 
Tbh, we implemented it differently, using threads and passing a function to each as well as the needed parameters and joining the threads in the end. But your way seems valid, in the end the performance evaluation by the teaching team will show which is faster 😎 

According to ChatGPT the reason for using futures is to not block the main thread. Was this also your reason to implement it this way? Do you know that there will a performance improvement because of this?

File: src/lib/storage/fixed_width_integer_vector.cpp (hunk 16 above)
Comment (phkeese) [reply src/lib/storage/fixed_width_integer_vector.cpp]: Could you please specify what you mean by "information it shouldn't have"? We chose to use the id `0` to represent NULL-values to simplify other code.
Thank you for the suggestion of `bit_width`.

Edit: No longer the case.

File: src/lib/storage/chunk.cpp (hunk 1 above)
Comment (ClFeSc) [reply src/lib/storage/chunk.cpp:13]: We evaluated the other options and found this one to be the cleanest solution. Especially regarding the solution presented in the lecture, we did not like the presence of an array of the types, that we would have to maintain separately. Also, understanding the solution from the lecture is not that easy.

File: src/lib/storage/dictionary_segment.cpp (hunk 5 above)
Comment (ClFeSc) [reply src/lib/storage/dictionary_segment.cpp:33]: We prefer the style of [returning early](https://medium.com/swlh/return-early-pattern-3d18a41bba8). This also applies to similar comments further below.

File: src/lib/storage/dictionary_segment.cpp (hunk 14 above)
Comment (ClFeSc) [reply src/lib/storage/dictionary_segment.cpp]: Unfortunately, the attribute vector does not have a capacity member.

File: src/lib/storage/fixed_width_integer_vector.cpp (hunk 15 above)
Comment (ClFeSc) [reply src/lib/storage/fixed_width_integer_vector.cpp]: This function is not a member of the `FixedWidthAttributeVector`. We thought it would fit here.
